std/algorithm/sorting.d

```diff
@@ -1048,147 +1048,287 @@ unittest
     {
-        this._r = rs;
-        adjustPosition();
+        final switch (_lastFrontIndex)
+        {
+            foreach (i, _; Rs)
+            {
+            case i:
+                debug Unqual!ElementType previousFront = source[i].front();
+                source[i].popFront();
+                debug if (!source[i].empty)
+                {
+                    assert(previousFront == source[i].front ||
+                           comp(previousFront, source[i].front),
+                           "Input " ~ i.stringof ~ " is unsorted"); // @nogc
+                }
+                break;
+            }
+        }
+        _lastFrontIndex = frontIndex;
     }
 
-    ///
-    @property bool empty()
+    static if (isBidirectional)
     {
-        return _crt == _crt.max;
-    }
+        @property auto ref back()
+        {
+            final switch (_lastBackIndex)
+            {
+                foreach (i, _; Rs)
+                {
+                case i:
+                    assert(!source[i].empty);
+                    return source[i].back;
+                }
+            }
+            assert(false);
+        }
 
-    ///
-    void popFront()
-    {
-        // Assumes _crt is correct
-        assert(!empty);
-        foreach (i, U; Rs)
+        private size_t backIndex()
         {
-            if (i < _crt) continue;
-            // found _crt
-            assert(!_r[i].empty);
-            _r[i].popFront();
-            adjustPosition();
-            return;
+            size_t bestIndex = size_t.max; // indicate undefined
+            Unqual!ElementType bestElement;
+            foreach (i, _; Rs)
+            {
+                if (!source[i].empty)
+                {
+                    if (bestIndex == size_t.max || // either this is the first or
+                        comp(bestElement, source[i].back))
+                    {
+                        bestIndex = i;
+                        bestElement = source[i].back;
+                    }
+                }
+            }
+            return bestIndex;
         }
-        assert(false);
-    }
 
-    ///
-    @property auto ref ElementType front()
-    {
-        assert(!empty);
-        // Assume _crt is correct
-        foreach (i, U; Rs)
+        void popBack()
         {
-            if (i < _crt) continue;
-            assert(!_r[i].empty);
-            return _r[i].front;
+            final switch (_lastBackIndex)
+            {
+                foreach (i, _; Rs)
+                {
+                case i:
+                    debug Unqual!ElementType previousBack = source[i].back();
+                    source[i].popBack();
+                    debug if (!source[i].empty)
+                    {
+                        assert(previousBack == source[i].back ||
+                               comp(source[i].front, previousBack),
+                               "Input " ~ i.stringof ~ " is unsorted"); // @nogc
+                    }
+                    break;
```

Comment:
These `foreach`es are static ones. I'm pretty sure that the break associates with `switch` here. If it we're I don't see how the unittests could pass.